app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+// @flow
+// import Utils from "libs/utils";
+
+const defaultItemsPerBatch = 10000;
+let idCounter = 0;
+
+// Some properties:
+// Insertion order is not guaranteed
+
+class DiffableMap<K: number, V> {
+  id: number;
+  chunks: Array<Map<K, V>>;
+  entryCount: number;
+  existsCache: Map<K, boolean>;
+  itemsPerBatch: number;
+
+  constructor(optKeyValueArray?: ?Array<[K, V]>, itemsPerBatch: number = defaultItemsPerBatch) {
+    // Make the id property not enumerable so that it does not interfere with tests
+    Object.defineProperty(this, "id", { value: idCounter++, writable: true });
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
+          newMap.chunks[idx] = new Map(this.chunks[idx]);
+          newMap.chunks[idx].set(key, value);
+          return newMap;
+        }
+        idx++;
+      }
+      throw new Error("should not happen");
+    } else {
+      const isTooFull = this.entryCount / this.chunks.length > this.itemsPerBatch;
+      const nonFullMapIdx =
+        isTooFull || this.chunks.length === 0 ? -1 : Math.floor(Math.random() * this.chunks.length);
```

Comment:
Doesn't `Math.random()` create conflicting indices?

Reply:
Why the boolean operator for generating a number?

Reply:
> Doesn't Math.random() create conflicting indices?

For each index, there is a chunk which typically holds multiple values. So, there is no "conflict". The random factor fills all chunks evenly (statistically).

> Why the boolean operator for generating a number?

> I'm also confused by what happens when isTooFull is true, which makes nonFullMapIdx=true, which then later yields to newDiffableMap.chunks[true] = new Map(this.chunks[true]);?

`||` binds stronger than the ternary operator. So, the index is set to `-1` iff the chunks are too full OR when there are no chunks. In this case, a new chunk is created.

Reply:
But `true || -1` yields `true` though due to short-circuit evaluation, doesn't it?

Reply:
The precedence is equal to `(isTooFull || length === 0) ? -1 :
randomIndex`. So, the short circuiting is only in the condition for the
ternary operator. Or do we talk about different things?

Reply:
Oh I see, sorry I misread your previous comment, makes sense!
Does prettier remove those parentheses if you explicitly set them?

Reply:
> Oh I see, sorry I misread your previous comment, makes sense!
Does prettier remove those parentheses if you explicitly set them?

Good idea, but yes, prettier removes them :see_no_evil: